cmsplugin_articles_ai/models.py

```diff
@@ -0,0 +1,289 @@
+# -*- coding: utf-8 -*-
+try:
+    from urllib.parse import urlencode  # Python 3
+except ImportError:
+    from urllib import urlencode  # Python 2
+
+from cms.models import CMSPlugin
+from django.contrib.auth.models import User
+from django.core.urlresolvers import reverse
+from django.db import models
+from django.db.models import Count, Q
+from django.utils import timezone
+from django.utils.encoding import python_2_unicode_compatible
+from django.utils.translation import ugettext_lazy as _
+from djangocms_text_ckeditor.fields import HTMLField
+from enumfields import Enum, EnumIntegerField
+from filer.fields.file import FilerFileField
+from filer.fields.image import FilerImageField
+
+
+class TagQuerySet(models.QuerySet):
+
+    @property
+    def as_url_encoded(self):
+        """
+        Return queryset as URL encoded string that is compatible with
+        the article views.
+        """
+        return urlencode({
+            "filter_tags": self.values_list("pk", flat=True)
+        })
+
+
+@python_2_unicode_compatible
+class Tag(models.Model):
+    name = models.CharField(verbose_name=_("nimi"), max_length=200, unique=True)
+    objects = TagQuerySet.as_manager()
+
+    class Meta:
+        verbose_name = _("tagi")
+        verbose_name_plural = _("tagit")
+        ordering = ('name',)
+
+    def __str__(self):
+        return self.name
+
+
+class ArticleQuerySet(models.QuerySet):
+
+    def public(self):
+        """
+        Returns articles that are considered public.
+        """
+        now = timezone.now()
+        return self.filter(Q(published_from__lte=now) & Q(Q(published_until__gte=now) | Q(published_until=None)))
+
+    def _get_tag_pks(self, tags):
+        """
+        Return list of tag ids from the given iterable of tags unless
+        the list already contains ids.
+        """
+        if hasattr(tags, "values_list"):
+            return tags.values_list("pk", flat=True)
+        elif all(isinstance(item, int) for item in tags):
+            # Looks like list already contains just ids
+            return tags
+        else:
+            return [tag.pk for tag in tags]
+
+    def tag_filter(self, filter_mode, tags):
+        """
+        Filter queryset according to given filter mode and given tags.
+        """
+        articles = self.all()
+        if filter_mode == TagFilterMode.ANY.value:
```

Comment:
Hmm... The fact that you need to use `.value` means that `filter_mode` isn't actually an `Enum` value.
You could cast it at the top of the function (`filter_mode = TagFilterMode(filter_mode)`).